lib/FreeRTOS/freertos-plus-mqtt/agent_command_pool.c

```diff
@@ -0,0 +1,165 @@
+
+    taskENTER_CRITICAL();
+    {
+        if( !initializedSemaphore )
+        {
+            memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+            freeCommandStructMutex = tempMutex;
+            destroySemaphore = false;
+        }
+
+        initializedSemaphore = true;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Destroy the semaphore if it was not used. */
+    if( destroySemaphore )
+    {
+        vSemaphoreDelete( tempMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t blockTimeMs )
+{
+    Command_t * structToUse = NULL;
+    size_t i;
+
+    /* Check here so we do not enter a critical section every time. */
+    if( !initializedSemaphore )
+    {
+        initializePool();
+        configASSERT( freeCommandStructMutex ); /*_RB_ Create all objects here statically. */
+    }
+
+    /* Check counting semaphore has been created. */
+    if( freeCommandStructMutex != NULL )
+    {
+        /* If the semaphore count is not zero then a command context is available. */
+        if( xSemaphoreTake( freeCommandStructMutex, pdMS_TO_TICKS( blockTimeMs ) ) == pdPASS )
+        {
+            for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+            {
+                taskENTER_CRITICAL();
+                {
+                    /* If the commandType is NONE then the structure is not in use. */
+                    if( commandStructurePool[ i ].commandType == NONE )
+                    {
+                        LogDebug( ( "Removed Command Context %d from pool", ( int ) i ) );
+                        structToUse = &( commandStructurePool[ i ] );
+
+                        /* To show the struct is no longer available to be returned
+                         * by calls to Agent_ReleaseCommand(). */
+                        structToUse->commandType = !NONE;
+                        taskEXIT_CRITICAL();
+                        break;
+                    }
+                }
+                taskEXIT_CRITICAL();
+            }
+        }
+    }
+
+    return structToUse;
+}
+
+/*-----------------------------------------------------------*/
+
+bool Agent_ReleaseCommand( Command_t * pCommandToRelease )
+{
+    size_t i;
+    bool structReturned = false;
+
+    /* See if the structure being returned is actually from the pool. */
+    for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+    {
+        if( pCommandToRelease == &( commandStructurePool[ i ] ) )
+        {
+            taskENTER_CRITICAL();
+
+            /* Yes its from the pool.  Clearing it to zero not only removes the old
+             * data it also sets the structure's commandType parameter to NONE to
+             * mark the structure as free again. */
+            memset( ( void * ) pCommandToRelease, 0x00, sizeof( Command_t ) );
+            taskEXIT_CRITICAL();
```

Comment:
This doesn't need a critical section, if you arrange to zero the commandType element last, does it?  Do you really need to zero the other elements?  Shouldn't acquisition assign every element?

Reply:
That's true, but with a memset this code won't have to be changed if we add any members to the command struct later. I'd prefer to get rid of the semaphore and instead send/receive addresses to/from a queue, which would solve many of the problems with the current implementation. However, I wanted to save that for a later PR, as this one copies the current code largely as-is to a new file